### DISCUSSION TITLE
[Crossfading] Fade duration is capped by cross duration

### DISCUSSION THREAD

**Discussion by pozhiloy-enotik** [savonet/liquidsoap#3538] [User Support]
**Describe the bug**
For example, I've set a manual ```liq_fade_out``` of 10 sec for debugging purpose.
Unfortunatly the track always end with a 2sec or less crossfade.

I've tried with the smart crossfade and the normal one.

> I've set up a beta test station here : https://webradio.roflcopter.fr/public/webradio_roflcopter_(beta_test) It has single playlist with 4 sequential tracks :
> You may find here the current Liquidsoap configuration : https://wtf.roflcopter.fr/paste/?932fe50622c3c81b#DioCfu9z5bCXAfY4X9owTMddCv2bFwtooJnTdyRSVKze

**To Reproduce**
```php
def live_aware_crossfade(old, new) =

    cross.simple(old.source, new.source, fade_in=2.00, fade_out=10.00)

end

radio = cross(minimum=0., duration=3.00, live_aware_crossfade, radio)
 ```
 
**Expected behavior**
![image](https://github.com/savonet/liquidsoap/assets/49619526/1a0fcc32-f567-4434-a7d1-e4a328a1a593)

**Version details**

    OS: Ubuntu 22.04
    Version 2.2.2

**Common issues**
https://github.com/AzuraCast/AzuraCast/issues/6713#issuecomment-1803744883

https://github.com/AzuraCast/AzuraCast/issues/6713#issue-1978969881
https://github.com/AzuraCast/AzuraCast/discussions/6252#discussioncomment-6946381

**Comment by toots** [comment]
Hi @pozhiloy-enotik ! You should also set the `duration` to `10.` to make sure the crossfade has enough data buffered for the duration of the cross.

**Comment by pozhiloy-enotik** [comment]
> Hi @pozhiloy-enotik ! You should also set the `duration` to `10.` to make sure the crossfade has enough data buffered for the duration of the cross.

This leads to even more sound overlap(10s actually). What people need is only 3s of overlap

**Comment by toots** [comment]
I see. you will need something a little more finer grained then. The way the crossfade works, both end and beginning of tracks are buffered and passed down to compute the crossfaded source.

In your case, since fade ou and fade in time are so different, you need to delay the starting source. Something like this:
```liquidsoap
def live_aware_crossfade(old, new) =
  new = sequence([blank(duration=5.), fade.in(duration=2., new)])
  add(normalize=false, [fade.out(duration=5., old), new])
end

radio = cross(minimum=0., duration=5.00, live_aware_crossfade, radio)
```

**Comment by pozhiloy-enotik** [comment]
> I see. you will need something a little more finer grained then. The way the crossfade works, both end and beginning of tracks are buffered and passed down to compute the crossfaded source.
> 
> In your case, since fade ou and fade in time are so different, you need to delay the starting source. Something like this:
> 
> ```
> def live_aware_crossfade(old, new) =
>   new = sequence([blank(duration=5.), fade.in(duration=2., new)])
>   add(normalize=false, [fade.out(duration=5., old), new])
> end
> 
> radio = cross(minimum=0., duration=5.00, live_aware_crossfade, radio)
> ```

Here you go: 
```At line 173, char 41-61: Error 5: this value has type (source(_), ...) -> _ (inferred at line 167 char 53 - line 171 char 3) but it should be a subtype of (unit, ...) -> _ ```

**Reply by toots** [reply]
This will do:

```liquidsoap
def live_aware_crossfade(old, new) =
  new = sequence([blank(duration=4.), fade.in(duration=2., new.source)])
  add(normalize=false, [new, fade.out(duration=5., old.source)])
end

s = cross(minimum=0., duration=5.00, live_aware_crossfade, s)
```

**Comment by pozhiloy-enotik** [comment]
> Hi @pozhiloy-enotik ! You should also set the `duration` to `10.` to make sure the crossfade has enough data buffered for the duration of the cross.

@toots Any way of buffering more data leaving the same duration?
```cross.simple(old.source, new.source, fade_in=2.00, fade_out=10.00)```

**Comment by pozhiloy-enotik** [comment]
Convert back to issue pls scince Fade duration is still capped by cross duration and I do not think that is intended behaviour

**Reply by pozhiloy-enotik** [reply]
@toots

**Comment by Russsgithub** [comment]
Hi @pozhiloy-enotik ,

There's a great explanation of how the transition mechanisms work in the liquidsoap [book](http://www.liquidsoap.info/book/book.pdf) (chapter 6.4 / page 114 onwards) .

The duration is the length of the whole transition, the fade out of the old track and the fade in of the new track.

If you want to use the example you give in the initial post

```
def live_aware_crossfade(old, new) =

    cross.simple(old.source, new.source, fade_in=2.00, fade_out=10.00)

end

radio = cross(minimum=0., duration=3.00, live_aware_crossfade, radio)
```

You need the duration at 10 seconds, because you have a 10 second fade out, and then add 8 seconds of blank audio in front of the new track, so it doesn't play audio when the transition is triggered. If we alter toots's function to implement this it might look something like

```

def live_aware_crossfade(old, new) =
  new = sequence([blank(duration=8.), fade.in(duration=2., new.source)])
  add(normalize=false, [new, fade.out(duration=10., old.source)])
end

s = cross(minimum=0., duration=10.00, live_aware_crossfade, s)
```
Here we add (with sequence) 8 seconds of blank audio to the beginning of the new track and then a 2 second fade in (10 seconds). The new track has a 10 second fade in so by mixing the 2 (with add) we get a 10 second fade out of the old track and 8 seconds of silence followed by a 2 second fade in on the old track.

As i mentioned above the book describes all this better than i do , so it's well worth looking through.

I hope this helps you.

**Reply by rrolla** [reply]
An interesting technique to add (with sequence) 8 seconds of blank audio to the beginning of the new track, this I also need to try
